nominal/experimental/video_processing/resolution.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import annotations
+
+import dataclasses
+from typing import Literal, Union
+
+from typing_extensions import TypeAlias
+
+
+@dataclasses.dataclass(frozen=True)
+class VideoResolution:
+    resolution_width: int | None = None
+    """Width of the video, in pixels. Auto-inferred if None based on aspect ratio.
+    NOTE: MUST be divisible by 2.
+    """
+
+    resolution_height: int | None = None
+    """Height of the video, in pixels. Auto-inferred if None based on aspect ratio.
+    NOTE: MUST be divisible by 2.
+    """
+
+    allow_upscaling: bool = False
+    """If true, allow upscaling beyond original resolution (e.g. 1080p -> 4k)"""
+
+    def __post_init__(self) -> None:
+        """Validate that provided resolution is valid"""
+        if self.resolution_height is not None:
+            if self.resolution_height <= 0 or self.resolution_height % 2 != 0:
+                raise ValueError(
+                    f"Provided resolution height is invalid-- must be positive and even integer"
+                    f", received {self.resolution_height}"
+                )
+
+        if self.resolution_width is not None:
+            if self.resolution_width <= 0 or self.resolution_width % 2 != 0:
+                raise ValueError(
+                    f"Provided resolution width is invalid-- must be positive and even integer"
+                    f", received {self.resolution_width}"
+                )
+
+    def scale_factor(self) -> str:
+        """Output a video filter flag usable with Ffmpeg to rescale the resolution of a video."""
+        # If the user has not provided a width, auto-compute a width that keeps the existing
+        # aspect ratio while also ensuring that the width is divisible by 2 (required for h264)
+        width_str = "-2"
+        if self.resolution_width is not None:
+            if self.allow_upscaling:
+                width_str = str(self.resolution_width)
+            else:
+                width_str = f"'min({self.resolution_width}, iw)'"
+
+        # If the user has not provided a height, auto-compute a height that keeps the existing
+        # aspect ratio while also ensuring that the height is divisible by 2 (required for h264)
+        height_str = "-2"
+        if self.resolution_height is not None:
+            if self.allow_upscaling:
+                height_str = str(self.resolution_height)
+            else:
+                height_str = f"'min({self.resolution_height}, ih)'"
+
+        # Set scale to desired resolution, and set the Sample Aspect Ratio (SAR) to be 1:1,
+        # meaning that each pixel of the video presents as a square when viewing in a video player
+        return f"scale={width_str}:{height_str},setsar=1:1"
```

Comment:
```suggestion
        return f"scale={width_str}:{height_str},setsar=1/1"
```

I suspect `setsar=1:1` translates to `setsar=r=1:max=1`, which happens to give a ratio of 1, but is confusing.